devicetypes/smartthings/fidure-2nd-generation-thermostat.src/fidure-2nd-generation-thermostat.groovy

```diff
@@ -0,0 +1,810 @@
+	 Integer idx = (descMap?.result == "success")? 20 : 18 // read attr has success but report attr does not.  Both parse the same
+	 String hex = descMap.raw.substring(idx)
+	 Integer len = Integer.parseInt(hex.substring(0, 2), 16)
+	 StringBuilder output = new StringBuilder()
+	 for (int i = 2; i < Math.min(len*2+2, hex.length()); i+=2) {
+		String str = hex.substring(i, i+2)
+		char c = (char) Integer.parseInt(str, 16)
+		output.append(c)
+	 }
+	 
+	 descMap.attrLength = len
+	 descMap.value = (String) output
+	 return descMap
+
+}
+
+private Map[] parseReportAttributeMessage(String description) {
+//  POSSIBLE ST BUG: when there are multiple attributes in the same message read attr mangles the value.  so we need to parse the RAW from scratch
+//	e.g. description = "read attr - raw: 82C101000022050000420C413137333052545F48413132, dni: 82C1, endpoint: 01, cluster: 0000, size: 22, attrId: 0005, result: success, encoding: 42, value: 413137333052545F48413132"
+
+	Map descMap = (description - "read attr - ").split(",").inject([:]) { map, param ->
+		def nameAndValue = param.split(":")
+		map += [(nameAndValue[0].trim()):nameAndValue[1].trim()]
+	}
+    Integer type = Integer.parseInt("${descMap.encoding}", 16)
+	Integer length = DataType.getLength(type)
+
+	// POSSIBLE ST BUG: sometimes a Char String attribute like model identifier is read through the ST platform but ST does not (according to docs) process Char Strings
+	if (0x42 == Integer.parseInt("${descMap.encoding}", 16))
+	{
+		descMap = getCharStringAttr(descMap)
+		length = descMap.attrLength
+	}
+	
+	Map[] descMaps = []
+
+    if (descMap.value.length() > (length * 2)){
+		descMaps += parseMultipleAttributes(descMap.cluster, descMap.raw.substring(12))
+	}else {
+		descMaps += descMap
+	}
+
+	Map[] resultMaps = []
+	descMaps.each { map ->
+		resultMaps += processSingleAttributeMessge(map)
+    }
+	return resultMaps
+}
+
+private Map parseThermostatClusterAttr(Map descMap){
+
+		Map map = [:]
+
+		switch(descMap.attrId.toLowerCase())
+				{
+	            case "0000":
+	  						map.name = "temperature"
+	  						map.value = getTemperature(descMap.value)
+	  					break
+	            case "0005":
+	            if (descMap.encoding ==  "23")
+	            {
+	        	    map.name = "holdExpiary"
+	              	map.value = "${convertToTime(descMap.value).getTime()}"
+	  			}
+	            break
+				case "0007":
+				map.name = "coolingDemand"
+				map.value = Integer.parseInt("${descMap.value}", 16)
+				break
+				case "0008":
+				map.name = "heatingDemand"
+				map.value = Integer.parseInt("${descMap.value}", 16)
+				break
+				case "0010":
+				Integer encoding = Integer.parseInt("${descMap.encoding}",16)
+					if (encoding == 0x30) {
+					map.name ="fanMode"
+					map.value = getFanModeMap()[descMap.value]
+					}
+				break
+	  			case "0011":
+				 if (descMap.encoding ==  "29") {
+					map.name = "coolingSetpoint"
+	  				map.value = getTemperature(descMap.value)
```

Comment:
scale value and add unit, and I assume thermostatSetpoint needs to be updated when either or both of cooling/heating setpoint is received depending on thermostatMode

Reply:
yes. fixed.